test/libraries/encoders/SettlementEncoder.sol

```diff
@@ -0,0 +1,228 @@
+            for (uint256 i = 0; i < orderRefunds.filledAmounts.length; i++) {
+                bytes memory filledAmount = orderRefunds.filledAmounts[i];
+                if (filledAmount.length != GPv2Order.UID_LENGTH) {
+                    revert("Invalid order UID length");
+                }
+                state.refunds.filledAmounts.push(filledAmount);
+            }
+        }
+
+        if (orderRefunds.preSignatures.length > 0) {
+            for (uint256 i = 0; i < orderRefunds.preSignatures.length; i++) {
+                bytes memory preSignature = orderRefunds.preSignatures[i];
+                if (preSignature.length != GPv2Order.UID_LENGTH) {
+                    revert("Invalid order UID length");
+                }
+                state.refunds.preSignatures.push(preSignature);
+            }
+        }
+    }
+
+    /// @dev Encode the current state into a settlement
+    function encode(State storage state, GPv2Settlement settlement) internal returns (EncodedSettlement memory) {
+        return EncodedSettlement({
+            tokens: tokens(state),
+            clearingPrices: state.tokenRegistrySlot.tokenRegistry().clearingPrices(),
+            trades: state.trades,
+            interactions: interactions(state, address(settlement))
+        });
+    }
+
+    /// @dev Encode just the setup interactions and use an otherwise empty encoded settlement
+    function encode(State storage, GPv2Interaction.Data[] memory setupInteractions)
+        internal
+        pure
+        returns (EncodedSettlement memory encodedSettlement)
+    {
+        encodedSettlement.interactions[uint256(InteractionStage.INTRA)] = setupInteractions;
+    }
+
+    /// @dev Encode the order refunds into interactions
+    function encodeOrderRefunds(State storage state, address settlement)
+        private
+        view
+        returns (GPv2Interaction.Data[] memory refunds_)
+    {
+        uint256 numInteractions =
+            (state.refunds.filledAmounts.length > 0 ? 1 : 0) + (state.refunds.preSignatures.length > 0 ? 1 : 0);
+        refunds_ = new GPv2Interaction.Data[](numInteractions);
+
+        uint256 i = 0;
+        if (state.refunds.filledAmounts.length > 0) {
+            refunds_[i++] = refundFnEncoder(
+                settlement, GPv2Settlement.freeFilledAmountStorage.selector, state.refunds.filledAmounts
+            );
+        }
+
+        if (state.refunds.preSignatures.length > 0) {
+            refunds_[i] = refundFnEncoder(
+                settlement, GPv2Settlement.freePreSignatureStorage.selector, state.refunds.preSignatures
+            );
+        }
+    }
+
+    /// @dev Encode a refund function call
+    function refundFnEncoder(address settlement, bytes4 fn, bytes[] memory orderUids)
+        private
+        pure
+        returns (GPv2Interaction.Data memory)
+    {
+        return GPv2Interaction.Data({target: settlement, value: 0, callData: abi.encodeWithSelector(fn, orderUids)});
```

Comment:
```suggestion
        if (state.refunds.filledAmounts.length > 0) {
            refunds_[i++] = refundFnEncoder(
                settlement, GPv2Settlement.freeFilledAmountStorage, state.refunds.filledAmounts
            );
        }

        if (state.refunds.preSignatures.length > 0) {
            refunds_[i] = refundFnEncoder(
                settlement, GPv2Settlement.freePreSignatureStorage, state.refunds.preSignatures
            );
        }
    }

    /// @dev Encode a refund function call
    function refundFnEncoder(address settlement, function(bytes[] calldata) fn, bytes[] memory orderUids)
        private
        pure
        returns (GPv2Interaction.Data memory)
    {
        return GPv2Interaction.Data({target: settlement, value: 0, callData: abi.encodeCall(fn, (orderUids))});
```

Reply:
Weird. I'm sure I tried to compile and it worked before, I would never have proposed something so fancy without trying it out first and making sure I got the syntax right! :laughing: 
Anyway, it looks like there's even a [Solidity unit test](https://github.com/ethereum/solidity/blob/v0.8.26/test/libsolidity/syntaxTests/types/contractTypeType/members/assign_function_via_contract_name_to_var.sol) designed to error on our exact use case, so I suppose it's a no-go.